File: scripts/test-db-connections.js
#!/usr/bin/env node

const { execSync } = require('child_process');
const path = require('path');
const fs = require('fs');

// Colors for terminal output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m'
};

// Helper to print colored text
function print(text, color = '') {
  console.log(color + text + colors.reset);
}

// Parse connection string to extract host
function extractHost(connectionString) {
  const match = connectionString.match(/@([^\/]+)\//);
  return match ? match[1] : 'unknown';
}

// Test a database connection using psql
async function testConnection(envName, connectionString) {
  print(`\n🔍 Testing ${envName} database...`, colors.bright);
  
  try {
    // Extract and display host
    const host = extractHost(connectionString);
    print(`   Host: ${host}`, colors.cyan);
    
    // Mask password in display
    const displayString = connectionString.replace(/:npg_[^@]+@/, ':***@').substring(0, 80) + '...';
    print(`   Connection: ${displayString}`);
    
    // Test connection using Node.js postgres module
    const testScript = `
      const postgres = require('postgres');
      const sql = postgres('${connectionString}', {
        ssl: 'require',
        max: 1,
        idle_timeout: 20,
        connect_timeout: 10,
      });
      
      (async () => {
        try {
          const result = await sql\`SELECT current_database(), current_user, version()\`;
          console.log(JSON.stringify({
            success: true,
            database: result[0].current_database,
            user: result[0].current_user
          }));
          await sql.end();
        } catch (error) {
          console.log(JSON.stringify({
            success: false,
            error: error.message
          }));
          process.exit(1);
        }
      })();
    `;
    
    // Write temporary test file
    const tempFile = path.join(__dirname, `.test-${envName.toLowerCase()}.js`);
    fs.writeFileSync(tempFile, testScript);
    
    // Run test
    const result = execSync(`cd frontend && node ${tempFile}`, { 
      encoding: 'utf8',
      cwd: path.join(__dirname, '..')
    });
    
    // Clean up temp file
    fs.unlinkSync(tempFile);
    
    const data = JSON.parse(result);
    if (data.success) {
      print(`   ✅ Connected successfully!`, colors.green);
      print(`   📊 Database: ${data.database}`);
      print(`   👤 User: ${data.user}`);
      return { success: true, host, database: data.database };
    } else {
      throw new Error(data.error);
    }
    
  } catch (error) {
    print(`   ❌ Connection failed: ${error.message}`, colors.red);
    return { success: false, host: extractHost(connectionString), error: error.message };
  }
}

// Load environment variables from a file
function loadEnvFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const envVars = {};
    
    content.split('\n').forEach(line => {
      if (line && !line.startsWith('#') && line.includes('=')) {
        const [key, ...valueParts] = line.split('=');
        envVars[key.trim()] = valueParts.join('=').trim();
      }
    });
    
    return envVars;
  } catch (error) {
    return null;
  }
}

// Main test function
async function runTests() {
  print('🚀 DreamOps Database Connection Tests', colors.bright + colors.blue);
  print('=====================================');
  print('Testing database separation for local, staging, and production environments\n');
  
  const projectRoot = path.join(__dirname, '..');
  const environments = [
    { name: 'LOCAL', backend: 'backend/.env.local', frontend: 'frontend/.env.local' },
    { name: 'STAGING', backend: 'backend/.env.staging', frontend: 'frontend/.env.staging' },
    { name: 'PRODUCTION', backend: 'backend/.env.production', frontend: 'frontend/.env.production' }
  ];
  
  const results = [];
  const hosts = new Set();
  
  // Check if postgres module is available
  try {
    require.resolve('postgres');
  } catch (e) {
    print('❌ postgres module not found. Installing...', colors.yellow);
    execSync('cd frontend && npm install postgres', { stdio: 'inherit' });
  }
  
  for (const env of environments) {
    print(`\n📂 Checking ${env.name} configuration...`, colors.bright);
    
    // Check backend file
    const backendPath = path.join(projectRoot, env.backend);
    const backendEnv = loadEnvFile(backendPath);
    
    if (!backendEnv) {
      print(`   ❌ Backend: ${env.backend} not found`, colors.red);
    } else if (!backendEnv.DATABASE_URL) {
      print(`   ❌ Backend: DATABASE_URL not configured`, colors.red);
    } else {
      print(`   ✅ Backend: ${env.backend} configured`, colors.green);
    }
    
    // Check frontend file
    const frontendPath = path.join(projectRoot, env.frontend);
    const frontendEnv = loadEnvFile(frontendPath);
    
    if (!frontendEnv) {
      print(`   ❌ Frontend: ${env.frontend} not found`, colors.red);
    } else if (!frontendEnv.POSTGRES_URL) {
      print(`   ❌ Frontend: POSTGRES_URL not configured`, colors.red);
    } else {
      print(`   ✅ Frontend: ${env.frontend} configured`, colors.green);
      
      // Test the connection
      const result = await testConnection(env.name, frontendEnv.POSTGRES_URL);
      results.push({ env: env.name, ...result });
      
      if (result.success) {
        hosts.add(result.host);
      }
    }
  }
  
  // Summary
  print('\n' + '='.repeat(50), colors.bright);
  print('📊 Test Summary', colors.bright);
  print('='.repeat(50));
  
  const successCount = results.filter(r => r.success).length;
  
  results.forEach(result => {
    if (result.success) {
      print(`✅ ${result.env}: Connected to ${result.host}`, colors.green);
    } else {
      print(`❌ ${result.env}: Failed - ${result.error || 'Not configured'}`, colors.red);
    }
  });
  
  // Database separation check
  print('\n' + '='.repeat(50), colors.bright);
  print('🔒 Database Separation Check', colors.bright);
  print('='.repeat(50));
  
  if (successCount === 3) {
    if (hosts.size === 3) {
      print('✅ Perfect! All 3 environments use different database hosts.', colors.green);
      print('   Your environments are properly isolated.', colors.green);
      
      // Show unique hosts
      print('\n📍 Unique Database Hosts:', colors.cyan);
      Array.from(hosts).forEach((host, i) => {
        print(`   ${i + 1}. ${host}`);
      });
    } else if (hosts.size === 2) {
      print('⚠️  Warning: Only 2 unique hosts detected!', colors.yellow);
      print('   Some environments might be sharing a database.', colors.yellow);
    } else {
      print('❌ Critical: All environments use the same host!', colors.red);
      print('   Please ensure you have separate Neon projects for each environment.', colors.red);
    }
  } else {
    print(`⚠️  Only ${successCount} of 3 connections succeeded.`, colors.yellow);
    print('   Please check the failed connections above.', colors.yellow);
  }
  
  // Additional info
  if (successCount === 3 && hosts.size === 3) {
    print('\n✨ Next Steps:', colors.cyan);
    print('1. Run database migrations for each environment');
    print('2. Set up GitHub Actions secrets for staging/production');
    print('3. Update deployment scripts to use the correct environment');
  }
}

// Run the tests
runTests().catch(error => {
  print(`\n❌ Test failed: ${error.message}`, colors.red);
  process.exit(1);
});

File: backend/test_db_connections.py
#!/usr/bin/env python3
"""
DreamOps Database Connection Tester
Tests that all three database environments are properly separated
"""

import asyncio
import os
import sys
from datetime import datetime
from pathlib import Path
import asyncpg
from dotenv import load_dotenv
import urllib.parse

# Colors for terminal output
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'

def print_colored(text, color):
    print(f"{color}{text}{Colors.ENDC}")

def parse_postgres_url(url):
    """Parse a PostgreSQL URL into components"""
    parsed = urllib.parse.urlparse(url)
    return {
        'host': parsed.hostname,
        'port': parsed.port or 5432,
        'user': parsed.username,
        'password': parsed.password,
        'database': parsed.path.lstrip('/'),
        'ssl': 'require' if 'sslmode=require' in url else 'prefer'
    }

async def test_connection(env_name, env_file):
    """Test database connection for a specific environment"""
    print(f"\n🔍 Testing {Colors.BOLD}{env_name}{Colors.ENDC} database...")
    
    # Check if env file exists
    if not Path(env_file).exists():
        print_colored(f"   ❌ File not found: {env_file}", Colors.RED)
        print(f"   💡 Create this file with your {env_name} Neon connection string")
        return {'success': False, 'env': env_name, 'error': 'File not found'}
    
    # Load the specific env file
    load_dotenv(env_file, override=True)
    db_url = os.getenv('DATABASE_URL')
    
    if not db_url:
        print_colored(f"   ❌ No DATABASE_URL found in {env_file}", Colors.RED)
        return {'success': False, 'env': env_name, 'error': 'No DATABASE_URL'}
    
    # Mask password in output
    masked_url = db_url[:30] + '***' + db_url[-20:] if len(db_url) > 50 else db_url
    print(f"   Connection: {masked_url}")
    
    try:
        # Parse connection details
        conn_params = parse_postgres_url(db_url)
        
        # Connect to database
        conn = await asyncpg.connect(
            host=conn_params['host'],
            port=conn_params['port'],
            user=conn_params['user'],
            password=conn_params['password'],
            database=conn_params['database'],
            ssl=conn_params['ssl']
        )
        
        # Test 1: Get database info
        db_info = await conn.fetchrow("""
            SELECT current_database() as db_name, 
                   current_user as db_user,
                   version() as db_version
        """)
        
        print_colored(f"   ✅ Connected successfully!", Colors.GREEN)
        print(f"   📊 Database: {db_info['db_name']}")
        print(f"   👤 User: {db_info['db_user']}")
        
        # Test 2: Check existing tables
        tables = await conn.fetch("""
            SELECT tablename 
            FROM pg_tables 
            WHERE schemaname = 'public'
            ORDER BY tablename
        """)
        
        print(f"   📋 Tables found: {len(tables)}")
        if tables:
            for i, table in enumerate(tables[:5]):
                print(f"      - {table['tablename']}")
            if len(tables) > 5:
                print(f"      ... and {len(tables) - 5} more")
        
        # Test 3: Create environment-specific test data
        timestamp = int(datetime.now().timestamp())
        test_table = f"test_{env_name.lower()}_{timestamp}"
        test_id = f"{env_name}-{timestamp}-test"
        
        # Create test table
        await conn.execute(f"""
            CREATE TABLE {test_table} (
                id SERIAL PRIMARY KEY,
                environment VARCHAR(50),
                test_id VARCHAR(100),
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        print_colored(f"   ✅ Created test table: {test_table}", Colors.GREEN)
        
        # Insert test data
        await conn.execute(f"""
            INSERT INTO {test_table} (environment, test_id) 
            VALUES ($1, $2)
        """, env_name, test_id)
        print(f"   ✅ Inserted test data with ID: {test_id}")
        
        # Verify data
        data = await conn.fetchrow(f"SELECT * FROM {test_table}")
        print(f"   ✅ Verified data: env={data['environment']}, id={data['test_id']}")
        
        # Check for test tables from other environments
        other_test_tables = await conn.fetch("""
            SELECT tablename 
            FROM pg_tables 
            WHERE schemaname = 'public' 
            AND tablename LIKE 'test_%'
            AND tablename != $1
        """, test_table)
        
        if other_test_tables:
            print_colored(f"   ⚠️  Found {len(other_test_tables)} test tables from other runs", Colors.YELLOW)
            for table in other_test_tables[:3]:
                print(f"      - {table['tablename']}")
        
        # Cleanup
        await conn.execute(f"DROP TABLE {test_table}")
        print(f"   🧹 Cleaned up test table")
        
        # Close connection
        await conn.close()
        
        return {
            'success': True, 
            'env': env_name, 
            'database': db_info['db_name'],
            'host': conn_params['host']
        }
        
    except Exception as e:
        print_colored(f"   ❌ Error: {str(e)}", Colors.RED)
        return {'success': False, 'env': env_name, 'error': str(e)}

async def cleanup_old_test_tables(env_file):
    """Clean up any old test tables from previous runs"""
    load_dotenv(env_file, override=True)
    db_url = os.getenv('DATABASE_URL')
    
    if not db_url:
        return
    
    try:
        conn_params = parse_postgres_url(db_url)
        conn = await asyncpg.connect(
            host=conn_params['host'],
            port=conn_params['port'],
            user=conn_params['user'],
            password=conn_params['password'],
            database=conn_params['database'],
            ssl=conn_params['ssl']
        )
        
        # Find old test tables
        old_tables = await conn.fetch("""
            SELECT tablename 
            FROM pg_tables 
            WHERE schemaname = 'public' 
            AND tablename LIKE 'test_%'
        """)
        
        if old_tables:
            print(f"\n🧹 Cleaning up {len(old_tables)} old test tables...")
            for table in old_tables:
                try:
                    await conn.execute(f"DROP TABLE IF EXISTS {table['tablename']}")
                    print(f"   Dropped: {table['tablename']}")
                except:
                    pass
        
        await conn.close()
    except:
        pass

async def main():
    """Run all connection tests"""
    print_colored("🚀 DreamOps Database Connection Tests", Colors.BOLD)
    print("=" * 50)
    print("This will verify that your 3 databases are separate and working.\n")
    
    # Change to backend directory
    backend_dir = Path(__file__).parent
    os.chdir(backend_dir)
    
    environments = [
        ("LOCAL", ".env.local"),
        ("STAGING", ".env.staging"),
        ("PRODUCTION", ".env.production")
    ]
    
    # Optional: cleanup old test tables
    if "--cleanup" in sys.argv:
        for env_name, env_file in environments:
            await cleanup_old_test_tables(env_file)
    
    results = []
    databases = set()
    hosts = set()
    
    for env_name, env_file in environments:
        result = await test_connection(env_name, env_file)
        results.append(result)
        
        if result['success']:
            databases.add(result['database'])
            hosts.add(result['host'])
    
    # Summary
    print(f"\n{Colors.BOLD}📊 Test Summary{Colors.ENDC}")
    print("=" * 50)
    
    for result in results:
        if result['success']:
            print_colored(f"✅ {result['env']}: Connected to {result['database']}", Colors.GREEN)
        else:
            print_colored(f"❌ {result['env']}: Failed - {result.get('error', 'Unknown error')}", Colors.RED)
    
    # Database separation check
    print(f"\n{Colors.BOLD}🔒 Database Separation Check{Colors.ENDC}")
    print("=" * 50)
    
    success_count = sum(1 for r in results if r['success'])
    
    if success_count == 3:
        if len(databases) == 3:
            print_colored("✅ Perfect! All 3 environments use different databases.", Colors.GREEN)
            print("   Your data is properly isolated between environments.")
        elif len(databases) == 2:
            print_colored("⚠️  Warning: Only 2 unique databases detected!", Colors.YELLOW)
            print("   Some environments might be sharing a database.")
        else:
            print_colored("❌ Critical: All environments use the same database!", Colors.RED)
            print("   This is dangerous - data will be mixed between environments!")
        
        if len(hosts) > 1:
            print(f"\n📍 Detected {len(hosts)} different database hosts")
    
    elif success_count > 0:
        print_colored(f"⚠️  Only {success_count} of 3 connections succeeded.", Colors.YELLOW)
        print("   Please check the failed connections above.")
    else:
        print_colored("❌ No database connections succeeded.", Colors.RED)
        print("\n💡 Setup Instructions:")
        print("1. Create 3 Neon projects at https://neon.tech")
        print("2. Copy each connection string")
        print("3. Create .env.local, .env.staging, and .env.production files")
        print("4. Add DATABASE_URL=<your-neon-connection-string> to each file")
        print("5. Run this test again")
    
    # Additional tips
    if success_count < 3:
        print("\n💡 Troubleshooting Tips:")
        print("- Make sure your .env files exist in the backend directory")
        print("- Check that DATABASE_URL is set in each file")
        print("- Verify your Neon connection strings are correct")
        print("- Ensure your IP is allowed (Neon allows all IPs by default)")

if __name__ == "__main__":
    asyncio.run(main())